core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -428,6 +428,68 @@ public MapIterator destructiveIterator() {
     return new MapIterator(numValues, new Location(), true);
   }
 
+  /**
+   * Iterator for the entries of this map. This is to first iterate over key index array
+   * `longArray` then accessing values in `dataPages`. NOTE: this is different from `MapIterator`
+   * in the sense that key index is preserved here
+   * (See `UnsafeHashedRelation` for example of usage).
+   */
+  public final class MapIteratorWithKeyIndex implements Iterator<Location> {
```

Comment:
Looks like `keyIndex` is not exposed outside this map iterator? then maybe call it `MapIteratorPreserveKeyIndex`?

Reply:
@viirya - you can think of `Location` returned by `MapIteratorWithKeyIndex.next()` indirectly exposes the `keyIndex`. I don't have a strong preference here, @cloud-fan WDYT here?